cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,999 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <algorithm>
+#include <execution>
+#include <optional>
+#include <raft/core/cuda_support.hpp>
```

Comment:
Sorry, Will, I happened to be looking through here more closely as I'm working on a PR to add the factory functions for creating the mdbuffer and I realized is that this file should have an `hpp`, not a `cuh` extension. `cuh` signifies that `nvcc` is needed to compile the file, whereas `hpp` extension might require cuda runtime, but doesn't need nvcc to compile. 

If the expectation that's being used here is that `cuh` signifies anything that requires `nvcc` or `cuda runtime` then we need to change that.

Reply:
Unfortunately, we need nvcc for [this guy right here](https://github.com/rapidsai/raft/pull/1999/files/fca74aa18c0a21ab9a689e9c26b6f582edbd3528#diff-7684a375b65e461c3fe89da16b806c1a9c9169cadf03bb25a0eff104a600fad2R38). That includes a kernel definition for some of the copies that mdbuffer requires.

Reply:
Sorry, that link doesn't seem to be taking me to the right line. Try [this](https://github.com/wphicks/raft/blob/fea-mdbuffer/cpp/include/raft/core/mdbuffer.cuh#L38) instead.

Reply:
Fair enough- yeah the copy does complicate things there and I admit I didn't notice that the first time around. I think what really stinks, though, that the copy itself forces device functions in the code for what could otherwise often just be plain cuda symbols (for eg. we are most often using standard strided layouts in the mdspans/mdarrays these days so the standard `cudaMemcpy` would get the job done most of the time). 

Anyways, you already have an `hpp` file for `mdbuffer` and I think we can (and should) revisit the potential for making the device-function-based copy an as-needed option at some point to avoid the propagation of the device functions. It's also not immediately obvious to me yet what that would look like.

Reply:
The closer I look, the more I realize we are getting inconsistent with this.. 

Take [math.hpp](https://github.com/wphicks/raft/blob/fea-mdbuffer/cpp/include/raft/core/math.hpp), which marks functions as device when compiling with `nvcc`. 

And also take [operators.hpp](https://github.com/wphicks/raft/blob/fea-mdbuffer/cpp/include/raft/core/operators.cuh) which includes the device functions so that they aren't included in `operators.hpp` (though `operators.hpp` includes `math.hpp`! 

I'm going to create an issue to revisit this for sure- we should definitely strive to be consistent everywhere.

Reply:
Yeah, I agree entirely. I'm wondering if for headers that are "optionally cuh" we might benefit from having a consistent flag that enables or disables the components that require nvcc. That would need to be separate from the current RAFT_DISABLE_CUDA definition, but it would give us an easy way to compose these files without an absurd propagation of options. Anyway, something we can discuss further in the issue, but I'm absolutely behind you on finding a more consistent global solution.